File: day05/day5.rb
# frozen_string_literal: true

# day 5
class BoardingPassProcessor
  def initialize(file)
    @data = File.readlines(file, chomp: true).map do |boarding_pass|
      [boarding_pass[0..6], boarding_pass[7..]]
    end
  end

  ROWS = (0..127).to_a.freeze
  COLS = (0..7).to_a.freeze
  SEAT_CHART = ROWS.product(COLS).freeze

  def calculate_max_seat_id
    @data.reduce(0) do |max_id, boarding_pass|
      seat_coords = boarding_pass.map(&method(:find_position))
      seat_id = calculate_seat_id.call(seat_coords)
      [seat_id, max_id].max
    end
  end

  def find_my_seat_id
    taken_seats = @data.map do |boarding_pass|
      boarding_pass.map(&method(:find_position))
    end
    taken_ids = taken_seats.map(&calculate_seat_id)
    SEAT_CHART.reject { |seat| taken_seats.include?(seat) }
              .map(&calculate_seat_id)
              .find { |id| ([id + 1, id - 1] - taken_ids).empty? }
  end

  private

  def calculate_seat_id
    ->((row_pos, col_pos)) { (row_pos * 8) + col_pos }
  end

  def find_position(instructions)
    orientation = instructions.size == 7 ? ROWS : COLS
    upper_char = orientation == ROWS ? 'F' : 'L'
    instructions.chars.reduce(orientation) do |pos, character|
      midpoint = pos.size / 2
      character == upper_char ? pos[0...midpoint] : pos[midpoint..]
    end.first
  end
end

processor = BoardingPassProcessor.new(*ARGV)

p processor.calculate_max_seat_id # -> 955
processor.find_my_seat_id # -> 569

File: day05/day5.js
const fs = require('fs');
const readFile = (path) => fs.readFileSync(path, { encoding: 'utf8' });

const BoardingPassProcessor = (file) => {
  const data = readFile(file).split('\n')
                             .slice(0, -1)
                             .map((pass) => [pass.slice(0, 7), pass.slice(7)]);

  const buildRange = (start, end) =>
    Array.from({ length: end - start + 1 }, (_, i) => start + i);
  const rows = buildRange(0, 127);
  const cols = buildRange(0, 7);
  const allSeats = () => {
    const seatCollection = [];
    for(const row of rows) {
      for(const col of cols) {
        seatCollection.push([row, col]);
      }
    }
    return seatCollection;
  }

  const calculateMaxSeatID = () => {
    return data.reduce((maxID, pass) => {
      const seatCoords = pass.map(findPosition)
      const seatID = calculateSeatID(...seatCoords);
      return Math.max(maxID, seatID)
    }, 0)
  }

  const findMySeat = () => {
    const allSeatIDs = allSeats().map((seat) => calculateSeatID(...seat));
    const takenIDs = data.map((pass) => {
      const seatCoords = pass.map(findPosition)
      return seatID = calculateSeatID(...seatCoords);
    })
    const vacantIDs = allSeatIDs.filter((id) => !takenIDs.includes(id));
    return vacantIDs.filter((id) =>
      takenIDs.includes(id + 1) && takenIDs.includes(id - 1));
  }

  const findPosition = (instructions) => {
    const orientation = instructions.length === 7 ? rows : cols;
    const upperChar = orientation === rows ? 'F' : 'L';
    return instructions.split('').reduce((chart, char) => {
      const midpoint = chart.length / 2;
      return char === upperChar ? chart.slice(0, midpoint) : chart.slice(midpoint);
    }, orientation)[0]
  }

  const calculateSeatID = (rowPos, colPos) => (rowPos * 8) + colPos;

  return {
    get data() { return data },
    calculateMaxSeatID,
    findMySeat
  }
}

const processor = BoardingPassProcessor(process.argv[2]);

processor.calculateMaxSeatID(); // 955
console.log(processor.findMySeat()); // 569
